python/dgl/nn/pytorch/explain/pgexplainer.py

```diff
@@ -358,3 +356,227 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
+        ...             h = self.conv(g, h)
+        ...
+        ...         if embed:
+        ...             return h
+        ...
+        ...         with g.local_scope():
+        ...             g.ndata["h"] = h
+        ...             hg = 0
+        ...             for ntype in g.ntypes:
+        ...                 hg = hg + dgl.mean_nodes(g, "h", ntype=ntype)
+        ...             return self.fc(hg)
+
+        >>> # Load dataset
+        >>> input_dim = 5
+        >>> hidden_dim = 5
+        >>> num_classes = 2
+        >>> g = dgl.heterograph({("user", "plays", "game"): ([0, 1, 1, 2], [0, 0, 1, 1])})
+        >>> g.nodes["user"].data["h"] = th.randn(g.num_nodes("user"), input_dim)
+        >>> g.nodes["game"].data["h"] = th.randn(g.num_nodes("game"), input_dim)
+
+        >>> transform = dgl.transforms.AddReverse()
+        >>> g = transform(g)
+
+        >>> # define and train the model
+        >>> model = Model(input_dim, hidden_dim, num_classes, g.canonical_etypes)
+        >>> optimizer = th.optim.Adam(model.parameters())
+        >>> for epoch in range(10):
+        ...     logits = model(g, g.ndata["h"])
+        ...     loss = th.nn.functional.cross_entropy(logits, th.tensor([1]))
+        ...     optimizer.zero_grad()
+        ...     loss.backward()
+        ...     optimizer.step()
+
+        >>> # Initialize the explainer
+        >>> explainer = dgl.nn.HeteroPGExplainer(model, hidden_dim)
+
+        >>> # Train the explainer
+        >>> # Define explainer temperature parameter
+        >>> init_tmp, final_tmp = 5.0, 1.0
+        >>> optimizer_exp = th.optim.Adam(explainer.parameters(), lr=0.01)
+        >>> for epoch in range(20):
+        ...     tmp = float(init_tmp * np.power(final_tmp / init_tmp, epoch / 20))
+        ...     loss = explainer.train_step(g, g.ndata["h"], tmp)
+        ...     optimizer_exp.zero_grad()
+        ...     loss.backward()
+        ...     optimizer_exp.step()
+
+        >>> # Explain the graph
+        >>> feat = g.ndata.pop("h")
+        >>> probs, edge_mask = explainer.explain_graph(g, feat)
+        """
+        self.model = self.model.to(graph.device)
+        self.elayers = self.elayers.to(graph.device)
+
+        embed = self.model(graph, feat, embed=True, **kwargs)
+        embed_flat = torch.concat(list(embed.values()))
+
+        homo_graph = to_homogeneous(graph, store_type=True)
+
+        edge_idx = homo_graph.edges()
+
+        col, row = edge_idx
+        col_emb = embed_flat[col.long()]
+        row_emb = embed_flat[row.long()]
+        emb = torch.cat([col_emb, row_emb], dim=-1)
+        emb = self.elayers(emb)
+        values = emb.reshape(-1)
+
+        values = self.concrete_sample(values, beta=tmp, training=training)
+        self.sparse_mask_values = values
+
+        reverse_eids = homo_graph.edge_ids(row, col).long()
+        edge_mask = (values + values[reverse_eids]) / 2
+
+        self.clear_masks()
+        self.set_masks(homo_graph, edge_mask)
```

Comment:
Do you still need to call `self.clear_masks()` and `self.set_masks` inside `explain_graph`? Is it sufficient to call them inside `train_step`?

Reply:
I don't see the need for the `clear_masks` calls as well.

with respect to the changes related to an [above comment](https://github.com/dmlc/dgl/pull/5739#discussion_r1214815720), I can keep the final clear based on the `training` param.

Reply:
I also think the `clear_masks` call is only necessary when there is no training, so following changes from [convo. above](https://github.com/dmlc/dgl/pull/5739#discussion_r1214815720) I think it should suffice to clear the mask when it is not in training setting

Reply:
I don't personally see any cases where `set_masks` is necessary inside the class because the steps from `explain_graph` shouldn't yeild a `None` for `edge_mask`?, but don't want to be too eager about pruning code.